### PR TITLE
Improve Pod Util config defaulting and validate all decorated jobs.

### PR DESCRIPTION
**Pull request by cjwagner**
Pod utility configuration did not properly merge the job specific values with the defaults, leaf fields in `UtilityConfig` and `GCSConfiguration` were not defaulted if the leaf struct was partially specified.
I also noticed that
- we were only validating decoration config for presubmit jobs.
- GCSConfiguration validation happened only at `gcsupload` runtime instead of when config is parsed.

/kind bug
/cc @amwat @krzyzacy @fejta 
/assign @stevekuznetsov

### CODE REVIEW

File: prow/apis/prowjobs/v1/types.go [hunk 1]
```diff
@@ -174,7 +174,7 @@ type DecorationConfig struct {
 	SSHHostFingerprints []string `json:"ssh_host_fingerprints,omitempty"`
 	// SkipCloning determines if we should clone source code in the
 	// initcontainers for jobs that specify refs
-	SkipCloning bool `json:"skip_cloning,omitempty"`
+	SkipCloning *bool `json:"skip_cloning,omitempty"`
```
Comment (cjwagner) [inline prow/apis/prowjobs/v1/types.go]: This change should behave properly with existing ProwJobs.

File: prow/config/config.go [hunk 2]
```diff
@@ -970,16 +976,55 @@ func setDecorationDefaults(provided, defaults *kube.DecorationConfig) *kube.Deco
 	}
 	if merged.UtilityImages == nil {
 		merged.UtilityImages = defaults.UtilityImages
+	} else if defaults.UtilityImages != nil {
```
[the remark this answers is not in the document]
Comment (cjwagner) [reply prow/config/config.go]: I moved all the defaulting and validation logic to the package where the structs are defined.

File: prow/config/config.go [hunk 3]
```diff
@@ -1029,10 +1074,28 @@ func validateDecoration(name string, spec *v1.PodSpec, config *kube.DecorationCo
 	if config.UtilityImages == nil {
 		return fmt.Errorf("job %s does not configure pod utility images but asks for decoration", name)
 	}
+	var missing string
```
[the remark this answers is not in the document]
Comment (cjwagner) [reply prow/config/config.go]: Done.

File: prow/apis/prowjobs/v1/types.go [hunk 4]
```diff
@@ -174,12 +175,81 @@ type DecorationConfig struct {
...
 	// CookieFileSecret is the name of a kubernetes secret that contains
 	// a git http.cookiefile, which should be used during the cloning process.
 	CookiefileSecret string `json:"cookiefile_secret,omitempty"`
 }
 
+func (d *DecorationConfig) ApplyDefault(def *DecorationConfig) *DecorationConfig {
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types.go]: note: this is beginning to look like an interface :)

File: prow/apis/prowjobs/v1/types_test.go [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+		GCSCredentialsSecret: "secretName",
+		SSHKeySecrets:        []string{"first", "second"},
+		SSHHostFingerprints:  []string{"primero", "segundo"},
+		SkipCloning:          &truth,
+	}
+
+	var testCases = []struct {
+		name     string
+		provided *DecorationConfig
+		expected *DecorationConfig
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types_test.go]: As exciting as copy-paste inheritance may be, `expected func(left, right *DecorationConfig) *DecorationConfig` will make these tests much easier to read

File: prow/apis/prowjobs/v1/types_test.go [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+	}{
+		{
+			name:     "nothing provided",
+			provided: &DecorationConfig{},
+			expected: &DecorationConfig{
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types_test.go]: Contrast this with
```golang
expected: func(left, right *DecorationConfig) *DecorationConfig { return left }
```

File: prow/apis/prowjobs/v1/types_test.go [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+				SSHHostFingerprints:  []string{"primero", "segundo"},
+				SkipCloning:          &truth,
+			},
+		},
+		{
+			name: "timeout provided",
+			provided: &DecorationConfig{
+				Timeout: 10 * time.Minute,
+			},
+			expected: &DecorationConfig{
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types_test.go]: Contrast this with:

```golang
expected: func(left, right *DecorationConfig) *DecorationConfig {
  out := *left
  out.Timeout = right.Timeout
  return &out
}
```

or if you're careful to not pass the defaults reference:

```golang
expected: func(left, right *DecorationConfig) *DecorationConfig {
  left.Timeout = right.Timeout
  return left
}
```

This is much easier to read because it hides all the nonsense we don't care about

File: prow/apis/prowjobs/v1/types_test.go [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+				},
+				GCSCredentialsSecret: "secretName",
+				SSHKeySecrets:        []string{"first", "second"},
+				SSHHostFingerprints:  []string{"unique", "print"},
+				SkipCloning:          &truth,
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types_test.go]: Run tests in parallel

`t.Run(testCase.name, func(t *testing.T) {...`

File: prow/apis/prowjobs/v1/types_test.go [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+		if actual, expected := testCase.provided.ApplyDefault(defaults), testCase.expected; !reflect.DeepEqual(actual, expected) {
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types_test.go]: consider making test cases not have to worry about overriding the defaults

```golang
caseDefault := *defaults
if a, e := testCase.provided.ApplyDefault(&caseDefault)...
```

File: prow/apis/prowjobs/v1/types.go [hunk 10]
```diff
@@ -220,6 +313,42 @@ type GCSConfiguration struct {
 	DefaultRepo string `json:"default_repo,omitempty"`
 }
 
+func (g *GCSConfiguration) ApplyDefault(def *GCSConfiguration) *GCSConfiguration {
+	if g == nil {
+		return def
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types.go]: Given that these are used all over the place, I believe it will be much safer to clone def and return the pointer to the cloned object, rather than allowing plugin foo to overwrite the configuration here.

File: prow/config/config_test.go [hunk 11]
```diff
@@ -498,6 +283,8 @@ periodics:
 				GCSConfiguration: &kube.GCSConfiguration{
 					Bucket:       "explicit-bucket",
 					PathStrategy: kube.PathStrategyExplicit,
+					DefaultOrg:   "kubernetes",
```
Comment (fejta) [inline prow/config/config_test.go]: Why are these changes necessary?

File: prow/apis/prowjobs/v1/types_test.go (hunk 9 above)
Comment (cjwagner) [reply prow/apis/prowjobs/v1/types_test.go]: Done.

File: prow/apis/prowjobs/v1/types_test.go (hunk 8 above)
Comment (cjwagner) [reply prow/apis/prowjobs/v1/types_test.go]: Done.

File: prow/apis/prowjobs/v1/types_test.go (hunk 5 above)
Comment (cjwagner) [reply prow/apis/prowjobs/v1/types_test.go]: Done.

File: prow/apis/prowjobs/v1/types.go (hunk 10 above)
Comment (cjwagner) [reply prow/apis/prowjobs/v1/types.go]: Done.

**Comment by cjwagner** [comment]
PTAL
